core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java

```diff
@@ -0,0 +1,313 @@
+    }
+    executorService.execute(new Runnable() {
+      @Override
+      public void run() {
+        // Please note that it is safe to release the lock and read into the read ahead buffer
+        // because either of following two conditions will hold - 1. The active buffer has
+        // data available to read so the reader will not read from the read ahead buffer.
+        // 2. This is the first time read is called or the active buffer is exhausted,
+        // in that case the reader waits for this async read to complete.
+        // So there is no race condition in both the situations.
+        boolean handled = false;
+        int read = 0;
+        Exception exception = null;
+        try {
+          while (true) {
+            read = underlyingInputStream.read(arr);
+            if (0 != read) break;
+          }
+          handled = true;
+        } catch (Exception ex) {
+          exception = ex;
+        } finally {
+          stateChangeLock.lock();
+          if (read < 0 || (exception instanceof EOFException) ) {
+            endOfStream = true;
+          } else if (!handled) {
+            readAborted = true;
+            readException = exception != null ? exception: new Exception("Unknown exception in ReadAheadInputStream");
+          } else {
+            byteBuffer.limit(read);
+          }
+          readInProgress = false;
+          signalAsyncReadComplete();
+          stateChangeLock.unlock();
+        }
+      }
+    });
+  }
+
+
+  private void signalAsyncReadComplete() {
+    stateChangeLock.lock();
+    try {
+      asyncReadComplete.signalAll();
+    } finally {
+      stateChangeLock.unlock();
+    }
+  }
+
+  private void waitForAsyncReadComplete() throws IOException {
+    stateChangeLock.lock();
+    try {
+      while (readInProgress)
+      asyncReadComplete.await();
+    } catch (InterruptedException e) {
+      throw  new InterruptedIOException(e.getMessage());
+    } finally {
+      stateChangeLock.unlock();
+    }
+  }
+
+  @Override
+  public int read() throws IOException {
+    int val = read(oneByte.get(), 0, 1);
+    if (val == -1) {
+      return -1;
+    }
+    return oneByte.get()[0] & 0xFF;
+  }
+
+  @Override
+  public int read(byte[] b, int offset, int len) throws IOException {
+    if (offset < 0 || len < 0 || offset + len < 0 || offset + len > b.length) {
+      throw new IndexOutOfBoundsException();
+    }
+    if (len == 0) {
+      return 0;
+    }
+    stateChangeLock.lock();
+    try {
+      return readInternal(b, offset, len);
+    } finally {
+      stateChangeLock.unlock();
+    }
+  }
+
+  /**
+   * Internal read function which should be called only from read() api. The assumption is that
+   * the stateChangeLock is already acquired in the caller before calling this function.
+   */
+  private int readInternal(byte[] b, int offset, int len) throws IOException {
+    assert (stateChangeLock.isLocked());
+    if (!activeBuffer.hasRemaining()) {
```

Comment:
This seems wrong when `activeBuffer.hasRemaining() == false && readAheadBuffer.hasRemaining() == true`. This may happen when `read` after  `skip(size of activeBuffer)`.

Reply:
Hmm.. not really, because if we `skip(size of active buffer)` then we will flip the active buffer and read ahead buffer. So the active buffer will actually have remaining bytes.